Src/PerformanceCollector/Shared/Implementation/QuickPulse/QuickPulseCollectionStateManager.cs

```diff
@@ -0,0 +1,222 @@
+            this.onSubmitSamples = onSubmitSamples;
+            this.onReturnFailedSamples = onReturnFailedSamples;
+        }
+        
+        public bool IsCollectingData
+        {
+            get
+            {
+                return this.isCollectingData;
+            }
+
+            private set
+            {
+                if (value != this.isCollectingData)
+                {
+                    // state transition
+                    this.ResetLastSuccessful();
+                }
+
+                this.isCollectingData = value;
+            }
+        }
+
+        public TimeSpan UpdateState(string instrumentationKey)
+        {
+            if (string.IsNullOrWhiteSpace(instrumentationKey))
+            {
+                return this.timings.ServicePollingInterval;
+            }
+
+            if (this.firstStateUpdate)
+            {
+                this.ResetLastSuccessful();
+
+                this.firstStateUpdate = false;
+            }
+
+            if (this.IsCollectingData)
+            {
+                // we are currently collecting
+                IList<QuickPulseDataSample> dataSamplesToSubmit = this.onSubmitSamples();
+
+                if (!dataSamplesToSubmit.Any())
+                {
+                    // no samples to submit, do nothing
+                    return this.DetermineBackOffs();
+                }
+
+                bool? keepCollecting = this.serviceClient.SubmitSamples(dataSamplesToSubmit, instrumentationKey);
+
+                QuickPulseEventSource.Log.TroubleshootingMessageEvent(string.Format(CultureInfo.InvariantCulture, "Sample submitted. Response: '{0}'", keepCollecting));
+
+                switch (keepCollecting)
+                {
+                    case null:
+                        // the request has failed, so we need to return the samples back to the submitter
+                        this.onReturnFailedSamples(dataSamplesToSubmit);
+                        break;
+
+                    case true:
+                        // the service wants us to keep collecting
+                        break;
+
+                    case false:
+                        // the service wants us to stop collection
+                        this.onStopCollection();
+                        break;
+                }
+
+                this.lastSuccessfulSubmit = keepCollecting.HasValue ? this.timeProvider.UtcNow : this.lastSuccessfulSubmit;
+                this.IsCollectingData = keepCollecting ?? this.IsCollectingData;
+            }
+            else
+            {
+                // we are currently idle and pinging the service waiting for it to ask us to start collecting data
+                bool? startCollection = this.serviceClient.Ping(instrumentationKey, this.timeProvider.UtcNow);
+
+                QuickPulseEventSource.Log.TroubleshootingMessageEvent(string.Format(CultureInfo.InvariantCulture, "Ping sent. Response: '{0}'", startCollection));
```

Comment:
Having formatting in verbose traces is not recommended. You add unnecessary noise. Could you please add seperate method, move string in the method argument?